Bugfix: Modules with same original name wrongly being marked as unused

seafoamteal opened the pull request:
# Description

This resolves #5145.

Three changes were made:
1. When a module was aliased, the entity for the original name has to be placed in the `Shadowed` layer and not the `Module` layer. This prevents it from shadowing previous imports with the same module name.
2. When registering a module reference from a module access site, the `module_name_to_node` map should not be checked. This map contains full original names of modules (e.g. `wibble/wobble` and `wobble`) and hence isn't guaranteed to point us to the correct node when we have limited information (e.g. just `wobble`). We skip this check and directly look up the entity in the Module layer. The `module_name_to_node` map is only needed/used when registering references from unqualified imports and aliases, since we are guaranteed to have information only about the original name of the module in those cases.
3. This is not directly related to the bugfix, but the `assert_no_warnings!` macro was defined ambiguously. If `src` was any `expr`, then the compiler would not know whether a second module tuple was part of the repeated list of module tuples or `src`. Changing `src` to be a `literal` fixes this.

# Example
## Before

```gleam
import wibble/wobble
import wobble as wob

pub fn main() {
  wobble.wimble()
  wob.womble()
}
```

Here, at `wobble.wimble()`, a lookup in the `module_name_to_node` map would return the top-level `wobble` module that is aliased to `wob`. Hence, no reference to `wibble/wobble` is registered and it is incorrectly marked unused.

A less visible problem is that `wibble/wobble` is added to the `Module` layer as entity with name `wobble`. Then, on the next line, `wobble` is once again added to the `Module` layer with the same name, shadowing the first import.

## After

At `wobble.wimble()`, we look up the `wobble` entity in the `Module` layer instead. This gives us the correct `wibble/wobble` module.

The `Module` layer contains two entities: `wobble` and `wob`. The `Shadowed` layer contains one entity: `wobble`.

seafoamteal committed:
Add test to replicate bug

Also updated assert_no_warnings to avoid ambiguity in parsing
test cases with multiple extra modules

seafoamteal committed:
Fix #5145

1. Modules with an alias had to be marked as shadowed
2. When registering module references from callsites, we can't look up
the module_name_to_node map as we have limited information about the
module name

seafoamteal committed:
Update CHANGELOG.md

lpil commented on compiler-core/src/reference.rs:
```diff
@@ -403,14 +418,22 @@ impl ReferenceTracker {
     }
 
-    pub fn register_module_reference(&mut self, name: EcoString) {
+    pub fn register_module_reference_for_unqualified_imports_and_aliases(
```
I don't understand this new name. It seems to say something about the context in which it is used, but not name the function itself, and the meaning is unclear.

Could these two functions be given clearer names please, and documentation comments could be useful 🙏

seafoamteal replied:
You're right, sorry about that. I've given them both better names now, relating to the sort of module name they take as arguments.

lpil reviewed (commented):
Thank you! I've left some questions inline.

seafoamteal committed:
Make function names clearer and add doc comments

Removed unnecessary register_unaliased_module function
Renamed both register_module_reference_* functions to be clearer about
intention
Added documentation for register_module_reference_* functions

seafoamteal commented on compiler-core/src/reference.rs:
```diff
@@ -307,20 +312,30 @@ impl ReferenceTracker {
     }
 
-    pub fn register_module(
+    pub fn register_unaliased_module(
```
I believe I was trying to work around making `EntityLayer` public and changing all the call sites of `register_module`, but it _does_ seem cleaner to avoid the extra wrapper function. I've made the changes necessary to remove it.

seafoamteal commented:
Hey @lpil, I'm just bumping this because I'm not sure if the inline responses show up as notifications.

lpil reviewed (commented):
Nice! Thank you! I think this is nearly ready, I've left some small notes RE semantics.

seafoamteal commented on compiler-core/src/reference.rs:
```diff
@@ -403,14 +409,28 @@ impl ReferenceTracker {
+    /// imports) as that is the only time we know the full module name. Since
+    /// full module names are not necessarily the name used to refer to the
+    /// module, this should NOT be used when inferring module access.
+    pub fn register_module_reference_with_module_name(&mut self, module_name: EcoString) {
```
I don't think this name is quite as clear as the other one. I tried it out and reverted it for now as it was confusing to me while I was reading through the code. Would `register_module_reference` be alright instead?

seafoamteal commented on compiler-core/src/reference.rs:
```diff
@@ -311,16 +316,17 @@ impl ReferenceTracker {
         &mut self,
         used_name: EcoString,
         module_name: EcoString,
+        layer: EntityLayer,
```
Hmm... yeah, I think that's what I was trying to avoid in my first iteration where there was a wrapper function around it for registering unaliased modules. Would you like me to go back to that implementation?

seafoamteal commented:
Hi Louis!

> What comments did you mean, sorry?

The ones in your inline review that I've replied to ^.^
